### PR TITLE
[Qt] Implement the runtime style API

### PR DESCRIPTION
**Pull request by tmpsantos**
Based on #5551, but using the core datatype converters.

### CODE REVIEW

**Comment by tmpsantos** [comment]
:eyes: @brunoabinader

File: platform/qt/include/qmapboxgl.hpp [hunk 1]
```diff
@@ -174,6 +174,8 @@ class Q_DECL_EXPORT QMapboxGL : public QObject
 
     void removeAnnotation(QMapbox::AnnotationID);
 
+    void setLayoutProperty(const QString& layer, const QString& property, const QVariant& value);
```
Comment (brunoabinader) [inline platform/qt/include/qmapboxgl.hpp]: Let's use the Qt style on the header files e.g. put the `&` together with the param name.

File: platform/qt/src/qt_conversion.hpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        ++iter;
+    }
+
+    return {};
+}
+
+inline optional<bool> toBool(const QVariant& value) {
+    if (value.type() == QVariant::Bool) {
```
Comment (brunoabinader) [inline platform/qt/src/qt_conversion.hpp]: I think a ternary operator would suit fine here, but this is fine too.

File: platform/qt/src/qt_conversion.hpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        return value.toBool();
+    } else {
+        return {};
+    }
+}
+
+inline optional<float> toNumber(const QVariant& value) {
+    if (value.type() == QVariant::Double) {
```
Comment (brunoabinader) [inline platform/qt/src/qt_conversion.hpp]: Ditto.

File: platform/qt/src/qt_conversion.hpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        return value.toFloat();
+    } else {
+        return {};
+    }
+}
+
+inline optional<std::string> toString(const QVariant& value) {
+    if (value.type() == QVariant::String) {
```
Comment (brunoabinader) [inline platform/qt/src/qt_conversion.hpp]: Can we use `switch (value.type())` here instead?

File: platform/qt/src/qt_conversion.hpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        return value.toString().toStdString();
+    } else if (value.type() == QVariant::Color) {
+        return value.value<QColor>().name().toStdString();
+    } else {
+        return {};
+    }
+}
+
+inline optional<Value> toValue(const QVariant& value) {
```
Comment (brunoabinader) [inline platform/qt/src/qt_conversion.hpp]: Ditto.

File: platform/qt/src/http_request.cpp [hunk 6]
```diff
@@ -39,7 +39,8 @@ QNetworkRequest HTTPRequest::networkRequest() const
     req.setRawHeader("User-Agent", "MapboxGL/1.0 [Qt]");
 
     if (m_resource.priorEtag) {
-        req.setRawHeader("If-None-Match", QByteArray(m_resource.priorEtag->data(), m_resource.priorEtag->size()));
+        auto etag = m_resource.priorEtag;
```
Comment (brunoabinader) [inline platform/qt/src/http_request.cpp]: Can we use `const auto` here?

**Comment by brunoabinader** [comment]
Minor comments above, but this is awesome! :tada:

File: platform/qt/src/qt_conversion.hpp (hunk 2 above)
Comment (tmpsantos) [inline platform/qt/src/qt_conversion.hpp]: List initialization would have to be explicit :cry:

File: platform/qt/src/qt_geojson.hpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    rapidjson::GenericDocument<rapidjson::UTF8<>, rapidjson::CrtAllocator> d;
+    if (data.endsWith(char(0))) {
+        d.Parse<0>(value.toByteArray().data());
+    } else {
+        d.Parse<0>(value.toByteArray().constData());
+    }
+
+    // Needed to keep the error message alive
+    // when we go out of this scope.
+    static std::string error;
```
Comment (jfirebaugh) [inline platform/qt/src/qt_geojson.hpp]: Can you open a ticket to change `Error` so we can remove this?

Comment (tmpsantos) [inline platform/qt/src/qt_geojson.hpp]: https://github.com/mapbox/mapbox-gl-native/pull/5655

**Pull request merged**